File: ios/App/App/NativeRoundBridge.swift
import Capacitor
import Foundation

/*
 NativeRoundBridge is the single native boundary for active-round consumers.
 It does not interpret golf rules: JavaScript Marshal validates every command
 and publishes the portable CaddyWatchScene. A future Live Activity/Lock Screen
 surface shares this bridge rather than reconstructing round state.

 It also does not own any wearable transport itself. It:
   - receives a Capacitor call
   - validates required top-level input
   - forwards the sanitised payload to WearableCoordinator
   - resolves the Capacitor call
   - publishes incoming wearable events to JavaScript

 Apple Watch's WatchConnectivity behaviour lives in
 Wearables/Apple/AppleWatchTransport.swift; a Garmin transport is a sibling
 registered with the same coordinator. JavaScript does not need to learn
 which platform it is talking to for ordinary Scene publication.
*/
@objc(NativeRoundBridge)
public final class NativeRoundBridge: CAPPlugin, CAPBridgedPlugin, WearableCoordinatorDelegate {
    public let identifier = "NativeRoundBridge"
    public let jsName = "NativeRoundBridge"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "publishScene", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "acknowledgeCommand", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "publishWatchMap", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "publishWatchMapAsset", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "watchMapInventory", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "publishWatchPlayer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "watchPlayerInventory", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "watchState", returnType: CAPPluginReturnPromise)
    ]

    private let coordinator = WearableCoordinator()
    private let queue = DispatchQueue(label: "com.claritygolf.caddy.native-round-bridge")
    /* The last inventory each side reported. Only a hint: JavaScript uses it
       to skip re-sending a package the wrist already has, and sending
       everything again is always a correct fallback. */
    private var watchMapInventoryReport: [String: Any]?
    private var watchPlayerInventoryReport: [String: Any]?

    /* Capacitor bridges JavaScript null as NSNull, which WatchConnectivity
       refuses (WCErrorCodePayloadUnsupportedTypes). The wearable decoder treats
       an absent key exactly like an explicit null, so dropping them is lossless. */
    private static func withoutNulls(_ value: Any) -> Any? {
        if value is NSNull { return nil }
        if let dictionary = value as? [String: Any] {
            var out = [String: Any]()
            for (key, item) in dictionary {
                if let kept = withoutNulls(item) { out[key] = kept }
            }
            return out
        }
        if let array = value as? [Any] {
            return array.compactMap { withoutNulls($0) }
        }
        return value
    }

    public override func load() {
        coordinator.delegate = self
        coordinator.register(AppleWatchTransport())
        coordinator.activateAll()
    }

    @objc public func publishScene(_ call: CAPPluginCall) {
        guard let scene = call.getObject("scene") else {
            call.reject("A CaddyWatchScene is required")
            return
        }
        let payload = (Self.withoutNulls(scene) as? [String: Any]) ?? [:]
        coordinator.publishScene(payload) { published in
            if published {
                call.resolve(["published": true])
            } else {
                call.resolve(["published": false, "queuedForReconciliation": true])
            }
        }
    }

    // MARK: - Watch lite maps

    @objc public func publishWatchMap(_ call: CAPPluginCall) {
        guard let manifest = call.getObject("manifest") else {
            call.reject("A Watch map manifest is required")
            return
        }
        let payload = (Self.withoutNulls(manifest) as? [String: Any]) ?? [:]
        coordinator.publishMapManifest(payload) { published in
            call.resolve(["published": published])
        }
    }

    @objc public func publishWatchMapAsset(_ call: CAPPluginCall) {
        guard let courseKey = call.getString("courseKey"), !courseKey.isEmpty,
              let asset = call.getString("asset"), !asset.isEmpty,
              let base64 = call.getString("base64"),
              let bytes = Data(base64Encoded: base64), !bytes.isEmpty else {
            call.reject("A Watch map asset requires courseKey, asset and base64 bytes")
            return
        }
        /* The package version is a millisecond timestamp. It crosses the
           Capacitor bridge as a string so it cannot be rounded on the way, and
           stays a string in the transfer metadata, which only carries
           property-list types. */
        let version = call.getString("version") ?? call.getInt("version").map(String.init) ?? ""
        guard !version.isEmpty else {
            call.reject("A Watch map asset requires a package version")
            return
        }
        coordinator.publishMapAsset(courseKey: courseKey, version: version, asset: asset, bytes: bytes) { result in
            switch result {
            case .success(let sent):
                call.resolve(["sent": sent])
            case .failure(let error):
                call.reject("Watch map asset could not be queued: \(error.localizedDescription)")
            }
        }
    }

    @objc public func watchMapInventory(_ call: CAPPluginCall) {
        queue.async { [weak self] in
            call.resolve(["inventory": self?.watchMapInventoryReport as Any])
        }
    }

    // MARK: - Watch player snapshot

    @objc public func publishWatchPlayer(_ call: CAPPluginCall) {
        guard let player = call.getObject("player") else {
            call.reject("A Watch player snapshot is required")
            return
        }
        /* An omitted My Bubble offset is the whole point of the field (see
           Bubble Bible s8) and Capacitor bridges a JS null as NSNull, which
           makes the entire send throw WCErrorCodePayloadUnsupportedTypes.
           Stripping is lossless: the Watch reads a missing key as nil. */
        let payload = (Self.withoutNulls(player) as? [String: Any]) ?? [:]
        coordinator.publishPlayer(payload) { published in
            call.resolve(["published": published])
        }
    }

    @objc public func watchPlayerInventory(_ call: CAPPluginCall) {
        queue.async { [weak self] in
            call.resolve(["inventory": self?.watchPlayerInventoryReport as Any])
        }
    }

    // MARK: - Watch presence

    @objc public func watchState(_ call: CAPPluginCall) {
        call.resolve(coordinator.state())
    }

    /* This is the only authoritative acknowledgement path. Native transport
       does not infer success: JavaScript returns the result after Marshal has
       accepted or rejected the command. */
    @objc public func acknowledgeCommand(_ call: CAPPluginCall) {
        guard let acknowledgement = call.getObject("acknowledgement"),
              acknowledgement["commandId"] as? String != nil else {
            call.reject("A command acknowledgement is required")
            return
        }
        /* An accepted command acknowledges with `reason: null`, which
           Capacitor bridges as NSNull and WatchConnectivity refuses outright
           (WCErrorCodePayloadUnsupportedTypes) - on BOTH the live and the
           queued path, so the Watch never heard that its LOCK or hole change
           went through and kept the button reading busy. Same fix as
           publishScene: the Watch decoder treats an absent key as nil. */
        let payload = (Self.withoutNulls(acknowledgement) as? [String: Any]) ?? [:]
        coordinator.acknowledge(payload) {
            call.resolve()
        }
    }

    // MARK: - WearableCoordinatorDelegate

    func wearableCoordinator(_ coordinator: WearableCoordinator, didReceiveCommand command: [String: Any]) {
        // JavaScript applies the generic command through its deduplicating
        // CaddyWatchBridge. Durable Watch command outbox/retry is the next
        // adapter milestone, not silently simulated here.
        notifyListeners("watchCommand", data: ["command": command], retainUntilConsumed: true)
    }

    /* The Watch's count of the holes it holds. Kept for the delivery module's
       next errand and pushed to JavaScript now, so the phone's handover card
       counts the same holes the wrist does. */
    func wearableCoordinator(_ coordinator: WearableCoordinator, didReceiveMapInventory inventory: [String: Any]) {
        queue.async { [weak self] in self?.watchMapInventoryReport = inventory }
        notifyListeners("watchMapInventory", data: ["inventory": inventory])
    }

    /* The wrist's own answer to "which bag do you already hold". Kept for a
       delivery module that attaches late and pushed to JavaScript now, so an
       unchanged bag is never re-sent. */
    func wearableCoordinator(_ coordinator: WearableCoordinator, didReceivePlayerInventory inventory: [String: Any]) {
        queue.async { [weak self] in self?.watchPlayerInventoryReport = inventory }
        notifyListeners("watchPlayerInventory", data: ["inventory": inventory])
    }

    /* Whether there is a wrist to hand the round to. JavaScript puts the answer
       on the Scene so the phone's Send to Watch and the Watch's own status strip
       read one fact. Pushed on activation and on every pairing/reachability
       change, and answerable on demand via watchState for a bridge that
       attaches late. */
    func wearableCoordinatorStateDidChange(_ coordinator: WearableCoordinator) {
        notifyListeners("watchState", data: coordinator.state())
    }
}

File: android/app/src/main/java/com/claritygolf/caddy/NativeRoundBridge.java
package com.claritygolf.caddy;

import com.claritygolf.caddy.wearables.garmin.GarminDeviceStore;
import com.claritygolf.caddy.wearables.garmin.GarminTransport;
import com.claritygolf.caddy.wearables.garmin.JsonBridge;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.Map;

/*
 NativeRoundBridge is Android's half of the native boundary for active-round
 consumers — the Capacitor plugin JavaScript already calls uniformly on
 every platform (app/js/native-round-bridge.js, watch-map-delivery.js,
 watch-player-delivery.js). It does not interpret golf rules: JavaScript
 Marshal validates every command and publishes the portable CaddyWatchScene.

 Unlike iOS's NativeRoundBridge — which was refactored to sit above a
 WearableCoordinator that can hold both an AppleWatchTransport and a
 GarminTransport — this talks to GarminTransport DIRECTLY. Android's only
 wearable target is Garmin (see the Garmin-on-Android scope note this
 project's memory records): there is nothing for a coordinator to arbitrate
 between, so that indirection is not built here. If Android ever gains a
 second wearable target, introducing a coordinator at that point is a known
 move, not a redesign — see WearableCoordinator.swift for the shape to copy.

 This plugin does:
   - receive a Capacitor call
   - validate required top-level input
   - convert JSObject/JSArray to plain Map/List (JsonBridge) so GarminTransport
     never depends on Capacitor types
   - forward to GarminTransport
   - resolve the Capacitor call
   - publish incoming GarminTransport events to JavaScript

 GarminTransport itself is currently an inert, safe stub — see its own
 header comment and garmin/README.md. This plugin compiles and resolves
 every call correctly today; it simply cannot yet reach a real Garmin
 device, because the Connect IQ Mobile SDK for Android is not vendored in
 this repo.
*/
@CapacitorPlugin(name = "NativeRoundBridge")
public class NativeRoundBridge extends Plugin implements GarminTransport.Listener {

    // Must always equal garmin/manifest.xml's <iq:application id="...">
    // once that placeholder is replaced — this is what scopes a message to
    // Caddy specifically among any other Connect IQ apps a paired device
    // might have installed.
    /** The same app as garmin/manifest.xml's {@code <iq:application id="...">}.
     *  Undashed 32-hex, exactly as the manifest spells it — the Connect IQ
     *  SDK for Android takes the application id as a String, not a
     *  java.util.UUID, so no reformatting is wanted here. (iOS deliberately
     *  holds the DASHED form of these same 128 bits, because its
     *  UUID(uuidString:) rejects the undashed one — see
     *  NativeRoundBridge.swift. Do not "fix" the two to match by spelling.)
     *  VERIFIED 2026-09-20 against the 2.4.0 AAR: IQApp takes the id as a
     *  plain String ({@code IQApp(String)}, {@code getApplicationId()}) with
     *  no UUID parsing anywhere in the Android SDK, so the manifest's own
     *  undashed spelling is the right one to pass here. */
    private static final String CONNECT_IQ_APP_ID = "fac5991c01f348ddb7577553071ddd0f";

    private GarminTransport transport;

    /* The last inventory each side reported. Only a hint, exactly as on
       iOS: JavaScript uses it to skip re-sending a package/bag the wrist
       already has, and sending everything again is always a correct
       fallback. */
    private Map<String, Object> watchMapInventoryReport;
    private Map<String, Object> watchPlayerInventoryReport;

    @Override
    public void load() {
        GarminDeviceStore deviceStore = new GarminDeviceStore(getContext());
        transport = new GarminTransport(getContext(), deviceStore, CONNECT_IQ_APP_ID);
        transport.setListener(this);
        transport.activate();
    }

    /* Releases the Connect IQ SDK's binding to Garmin Connect Mobile. The
       chosen device is kept in GarminDeviceStore, so this costs a re-bind on
       next launch, never a re-pair. */
    @Override
    protected void handleOnDestroy() {
        if (transport != null) { transport.deactivate(); }
        super.handleOnDestroy();
    }

    @PluginMethod
    public void publishScene(PluginCall call) {
        JSObject scene = call.getObject("scene");
        if (scene == null) {
            call.reject("A CaddyWatchScene is required");
            return;
        }
        transport.publishScene(JsonBridge.toMap(scene), published -> {
            JSObject result = new JSObject();
            if (published) {
                result.put("published", true);
            } else {
                result.put("published", false);
                result.put("queuedForReconciliation", true);
            }
            call.resolve(result);
        });
    }

    @PluginMethod
    public void publishWatchMap(PluginCall call) {
        JSObject manifest = call.getObject("manifest");
        if (manifest == null) {
            call.reject("A Watch map manifest is required");
            return;
        }
        transport.publishMapManifest(JsonBridge.toMap(manifest), published -> {
            JSObject result = new JSObject();
            result.put("published", published);
            call.resolve(result);
        });
    }

    /* Android/Garmin never receives map imagery as pushed bytes — Garmin
       pulls each hole raster by URL (see garmin/GarminMapDownloader.mc's
       header comment: there is no public Monkey C API for decoding an
       arbitrary byte buffer the app assembled itself, but
       Communications.makeImageRequestWithDictionary decodes a fetched URL
       directly). This method still has to exist and resolve successfully,
       though: app/js/watch-map-delivery.js gates its ENTIRE native map
       delivery — including publishWatchMap, which Garmin does need for its
       manifest — on both publishWatchMap AND publishWatchMapAsset being
       present as functions. Omitting this method would silently break
       manifest delivery too, not just asset delivery. */
    @PluginMethod
    public void publishWatchMapAsset(PluginCall call) {
        String courseKey = call.getString("courseKey");
        String asset = call.getString("asset");
        String base64 = call.getString("base64");
        if (courseKey == null || courseKey.isEmpty() || asset == null || asset.isEmpty()
                || base64 == null || base64.isEmpty()) {
            call.reject("A Watch map asset requires courseKey, asset and base64 bytes");
            return;
        }
        JSObject result = new JSObject();
        result.put("sent", false);
        call.resolve(result);
    }

    @PluginMethod
    public void watchMapInventory(PluginCall call) {
        JSObject result = new JSObject();
        result.put("inventory", watchMapInventoryReport == null ? null : JsonBridge.toJSObject(watchMapInventoryReport));
        call.resolve(result);
    }

    @PluginMethod
    public void publishWatchPlayer(PluginCall call) {
        JSObject player = call.getObject("player");
        if (player == null) {
            call.reject("A Watch player snapshot is required");
            return;
        }
        transport.publishPlayer(JsonBridge.toMap(player), published -> {
            JSObject result = new JSObject();
            result.put("published", published);
            call.resolve(result);
        });
    }

    @PluginMethod
    public void watchPlayerInventory(PluginCall call) {
        JSObject result = new JSObject();
        result.put("inventory", watchPlayerInventoryReport == null ? null : JsonBridge.toJSObject(watchPlayerInventoryReport));
        call.resolve(result);
    }

    @PluginMethod
    public void watchState(PluginCall call) {
        call.resolve(stateAsJSObject());
    }

    // ------------------------------------ Garmin pairing (Settings > Garmin)

    @PluginMethod
    public void garminState(PluginCall call) {
        call.resolve(JsonBridge.toJSObject(transport.garminState()));
    }

    /** Resolves with { devices: [...], sdkLinked, reason }. An empty list with
     *  sdkLinked false means "we cannot look", which the settings page words
     *  differently from "we looked and found none". */
    @PluginMethod
    public void garminDevices(PluginCall call) {
        call.resolve(JsonBridge.toJSObject(transport.availableDevices()));
    }

    @PluginMethod
    public void selectGarminDevice(PluginCall call) {
        String deviceId = call.getString("deviceId");
        if (deviceId == null || deviceId.trim().isEmpty()) {
            call.reject("A deviceId is required");
            return;
        }
        transport.selectDevice(
            deviceId,
            call.getString("deviceName", ""),
            call.getString("model", "")
        );
        call.resolve(JsonBridge.toJSObject(transport.garminState()));
    }

    @PluginMethod
    public void clearGarminDevice(PluginCall call) {
        transport.clearSelectedDevice();
        call.resolve(JsonBridge.toJSObject(transport.garminState()));
    }

    /* The paid gate. JavaScript owns the membership question
       (ClarityPayments.hasActiveAccess) and pushes the answer down; the
       transport refuses to send while it is false. Defaults to false, so
       nothing reaches a Garmin until something affirmatively says it may. */
    @PluginMethod
    public void setGarminEnabled(PluginCall call) {
        transport.setEntitled(Boolean.TRUE.equals(call.getBoolean("enabled", Boolean.FALSE)));
        call.resolve();
    }

    /* This is the only authoritative acknowledgement path. Native transport
       does not infer success: JavaScript returns the result after Marshal
       has accepted or rejected the command. */
    @PluginMethod
    public void acknowledgeCommand(PluginCall call) {
        JSObject acknowledgement = call.getObject("acknowledgement");
        if (acknowledgement == null || acknowledgement.getString("commandId") == null) {
            call.reject("A command acknowledgement is required");
            return;
        }
        transport.acknowledge(JsonBridge.toMap(acknowledgement), () -> call.resolve());
    }

    // -------------------------------------------------- GarminTransport.Listener

    @Override
    public void onCommandReceived(Map<String, Object> command) {
        // JavaScript applies the generic command through its deduplicating
        // CaddyWatchBridge. Matches NativeRoundBridge.swift's watchCommand
        // event exactly, including retainUntilConsumed(true) — a command
        // that arrives before JavaScript has attached its listener must not
        // be dropped.
        JSObject data = new JSObject();
        data.put("command", JsonBridge.toJSObject(command));
        notifyListeners("watchCommand", data, true);
    }

    @Override
    public void onMapInventoryReceived(Map<String, Object> inventory) {
        watchMapInventoryReport = inventory;
        JSObject data = new JSObject();
        data.put("inventory", JsonBridge.toJSObject(inventory));
        notifyListeners("watchMapInventory", data, false);
    }

    @Override
    public void onPlayerInventoryReceived(Map<String, Object> inventory) {
        watchPlayerInventoryReport = inventory;
        JSObject data = new JSObject();
        data.put("inventory", JsonBridge.toJSObject(inventory));
        notifyListeners("watchPlayerInventory", data, false);
    }

    @Override
    public void onStateChanged() {
        notifyListeners("watchState", stateAsJSObject(), false);
    }

    private JSObject stateAsJSObject() {
        GarminTransport.State state = transport.state();
        JSObject result = new JSObject();
        result.put("supported", state.supported);
        result.put("activated", state.activated);
        result.put("paired", state.paired);
        result.put("appInstalled", state.appInstalled);
        result.put("reachable", state.reachable);
        return result;
    }
}
